alacritty_terminal/src/tty/windows/conpty.rs

```diff
@@ -79,63 +82,92 @@ impl ConptyApi {
     }
 }
 
-/// RAII Pseudoconsole
+/// RAII Pseudoconsole. Closes pipes and Pty when dropped.
 pub struct Conpty {
-    pub handle: HPCON,
-    api: ConptyApi,
+    inner: Arc<ConptyInner>,
+    pub conout: EventedReadablePipe,
+    pub conin: EventedWritablePipe,
 }
 
-/// Handle can be cloned freely and moved between threads.
-pub type ConptyHandle = Arc<Conpty>;
+impl Conpty {
+    pub fn new(size: &SizeInfo) -> Option<Self> {
+        let api = ConptyApi::new()?;
+
+        let mut handle = 0 as HPCON;
+
+        // Passing 0 as the size parameter allows the "system default" buffer
+        // size to be used. There may be small performance and memory advantages
+        // to be gained by tuning this in the future, but it's likely a reasonable
+        // start point.
+        let (conout, conout_pty_handle) = miow::pipe::anonymous(0).unwrap();
+        let (conin_pty_handle, conin) = miow::pipe::anonymous(0).unwrap();
+
+        let coord =
+            coord_from_sizeinfo(size).expect("Overflow when creating initial size on pseudoconsole");
+
+        // Create the Pseudo Console, using the pipes
+        let result = unsafe {
+            (api.CreatePseudoConsole)(
+                coord,
+                conin_pty_handle.into_raw_handle(),
+                conout_pty_handle.into_raw_handle(),
+                0,
+                &mut handle as *mut HPCON,
+            )
+        };
+
+        assert_eq!(result, S_OK);
+
+        let inner = ConptyInner { handle, api, closed: AtomicBool::new(false) };
+
+        Some(Self {
+            inner: Arc::new(inner),
+            conout: EventedReadablePipe::Anonymous(EventedAnonRead::new(conout)),
+            conin: EventedWritablePipe::Anonymous(EventedAnonWrite::new(conin)),
+        })
+    }
+
+    pub fn resize_handle(&self) -> ConptyResizeHandle {
+        ConptyResizeHandle(self.inner.clone())
+    }
+}
+
+/// Resize handle can be moved between threads and outlive the ConPTY safely
+pub struct ConptyResizeHandle(Arc<ConptyInner>);
+
+pub struct ConptyInner {
+    handle: HPCON,
+    api: ConptyApi,
+    closed: AtomicBool,
+}
 
 impl Drop for Conpty {
     fn drop(&mut self) {
-        unsafe { (self.api.ClosePseudoConsole)(self.handle) }
+        // Mark as closed so resize handles won't attempt to resize the closed pseudoconsole
+        self.inner.closed.store(true, Ordering::SeqCst);
+
+        // NB the i/o pipes must be closed (dropped) at the same time as the ClosePseudoConsole
+        // call. Else this call can deadlock!
+        unsafe { (self.inner.api.ClosePseudoConsole)(self.inner.handle) }
```

Comment:
We can't close two things at the same time, we either have to close them before or after. I'm not quite sure what the exact parameters of the restraints are just from this comment and the existing code.

Based on the code, it seems like the I/O pipes are automatically dropped once `Conpty`'s destructor is done? So *after* the `ClosePseudoConsole`?

I also don't think `NB` has ever been used as a comment type in the Alacritty repo and it seems generally kinda uncommon. For annotations about things to watch out for I've been using `XXX` so far I think, would be good to be consistent here.